store/sqlstore/upgrade.go

```diff
@@ -1208,6 +1209,201 @@ func upgradeDatabaseToVersion537(sqlStore *SqlStore) {
 	sqlStore.RemoveIndexIfExists("IDX_RetentionPolicies_DisplayName_Id", "RetentionPolicies")
 	sqlStore.CreateIndexIfNotExists("IDX_RetentionPolicies_DisplayName", "RetentionPolicies", "DisplayName")
 
+	fixCRTThreadCountsAndUnreads(sqlStore)
+	fixCRTChannelMembershipCounts(sqlStore)
+	fixCRTChannelUnreadsForJoinLeaveMessages(sqlStore)
+
 	// saveSchemaVersion(sqlStore, Version5370)
 	// }
 }
+
+// fixCRTThreadCountsAndUnreads
+// 1. Fixes reply counts, last reply times and participants of threads.
+// 2. Marks threads as read for users where the last reply time of the thread
+//    is earlier than the time the user viewed the channel.
+//    Marking a thread means setting the mention count to zero and
+//    setting the last viewed at time of the the thread as the last viewed at time
+//    of the channel
+func fixCRTThreadCountsAndUnreads(sqlStore *SqlStore) {
+	countsAndLastReplyCTE := `
+		SELECT RootId, COUNT(Id) AS FixedReplyCount, max(Posts.CreateAt) as FixedLastReply
+		FROM Posts where Posts.DeleteAt = 0
+		GROUP BY RootId
```

Comment:
That's one massive read, right? Counting all rows of posts for every thread? If there's millions of posts in the DB this might take a long time. Can we make it so we only count the replies for root ids that we care about instead of every single root id?

Reply:
Agree. Not even considering the query time, it has to scan the _entire_ posts table without any index and group it. This has the potential to cause an OOM.

I would recommend to process in batches. Use a cursor for the post id, and move the cursor every batch. That's much more efficient. It uses the PK index, and processes small batches in memory.

The `LastReplyAt` is a max field, so it shouldn't matter if a root_id group falls in 2 batches. The `ReplyCount` is a bit tricky. One way might be to set the field to zero initially, and then increment it in every batch.